CARDS-2332: Add automatic saving and checkin of forms after a period of inactivity

veronikaslc opened the pull request:
To test use the branch [CARDS-2332-test](https://github.com/data-team-uhn/cards/tree/CARDS-2332-test)

veronikaslc committed:
CARDS-2332: Add automatic saving and checkin of forms after a period of inactivity

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -113,6 +117,11 @@ function Form (props) {
 
+  let [ showCountdownModal, setShowCountdownModal ] = useState(false);
+  let [ showSessionExpiredAlert, setShowSessionExpiredAlert ] = useState(false);
+  const TWENTYSEVEN_MINS_IN_MS = 27 * 60 * 1000;
```
Let's call this something more generic, such as SESSION_EXPIRATION_DELAY

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -466,6 +511,15 @@ function Form (props) {
             </div>
   )
+  
+  let getExpiryMessage = () => {
+    let minutes = Math.floor((countDown % (1000 * 60 * 60)) / (1000 * 60));
+    let seconds = Math.floor((countDown % (1000 * 60)) / 1000);
+    let minutesMessage = minutes > 0 ? minutes == 1 ? '1 minute' : '2 minutes' : '';
+    let secondsMessage = minutes == 0 ? `${seconds} seconds` : ''; 
+    let message = `Your session will expire in ${minutesMessage}${secondsMessage}. You should save your answers now to keep your session active and prevent unsaved data loss.`;
+    return message;
```
This is very customized for a 2min interval. Let's do something more generic like in here:  https://github.com/data-team-uhn/cards/blob/dev/modules/patient-portal/src/main/frontend/src/patient-portal/QuestionnaireSet.jsx#L586-L610

veronikaslc replied:
That would be a 30 lines of code overcomplication for a simple case when we either display minutes or seconds, and minutes of values either 1 or 2.

marta- replied:
But it would have the advantage that we can easily change our minds if we want the counter to be on for 1 minute or 5.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -113,6 +117,11 @@ function Form (props) {
+  let [ showCountdownModal, setShowCountdownModal ] = useState(false);
+  let [ showSessionExpiredAlert, setShowSessionExpiredAlert ] = useState(false);
+  const TWENTYSEVEN_MINS_IN_MS = 27 * 60 * 1000;
+  let [ countDown, setCountDown ] = useState(TWENTYSEVEN_MINS_IN_MS + new Date().getTime());
```
I'm not understanding the meaning of countdown. We're initializing it with a time that is 27 minutes _from now_ (it includes current time, so it could be something like 10hours and 45 minutes), but on line 177 we're comparing it to 1second. Something doesn't fit.

veronikaslc replied:
We need to initialize it somewhere to that value so we can start counting down 27 minutes subtracting current time from it on line 181 to see if it's close to opening a dialog. That place is the right one for initialization since we mount the component.
Each time user saves the form we re-initialize it again the same way.
We can have a separate variable to hold the countdown start time, but we technically have no need in it.

marta- replied:
We don't need 2 variables. We only need countDown (initialized at 2min), and we need 2 timers: one in 27min (a constant we already have) that pops the dialog, and one that updates countDown that is displayed in the dialog.

veronikaslc replied:
You mean avoid involving current time usage at all and all calculations subtracting it?

marta- replied:
yes

veronikaslc replied:
But the specs directed using current timestamps of events

`when opening a form or when done saving, record in a variable the time when the save finished`

Isn't it the most natural way to deal with time?

marta- replied:
@sdumitriu what was the purpose of recording the save timestamp? If the functionality is in the Form, we already have the lastSaveTimestamp.

sdumitriu replied:
`lastSaveTimestamp` is what was needed, no need for a new variable.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -162,7 +171,41 @@ function Form (props) {
+    const interval = setInterval(() => {
+      let timeLeft = countDown - new Date().getTime();
+      if (timeLeft >= 0) {
+        setCountDown(timeLeft);
```
With this line the "countDown" changes meaning since we subtract the current time.

veronikaslc replied:
I'd say here the countDown is starting to act up to its meaning as we are actively counting down

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -162,7 +171,41 @@ function Form (props) {
+      } else {
+        clearInterval(interval);
+      }
+    }, 1000);
```
There's no need to do checks every second for 27 minutes. We can set the timer in 27 minutes to launch the checkin. To cancel upon save, `interval` should be a state variable (ideally with a better name, like `autoCheckinTimer`) that gets reset on save.

veronikaslc replied:
We still need the count down of every second for the dialog. Why do we need to introduce another timer to launch this timer?

marta- replied:
We need to count every second for a couple of minutes at the end. For 20+ minutes in the beginning, looping through code that doesn't really do anything is pointless.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -618,6 +673,40 @@ function Form (props) {
           <Typography variant="body1" paragraph>Time of the last successful save: {DateTime.fromISO(lastSaveTimestamp.toISOString()).toRelativeCalendar()}</Typography>
         }
       </ErrorDialog>
+      <ResponsiveDialog title="Session will expire soon" open={showCountdownModal} onClose={onClose}>
+        <DialogContent dividers>
+          <Typography variant="body1" paragraph>{getExpiryMessage()}</Typography>
+        </DialogContent>
+        <DialogActions>
+          <Button
+            onClick={() => {
+                setShowCountdownModal(false);
+                // trigger a normal save
+                // ...make sure that on paginated forms this doesn’t go to the next page
+                const cashedRequireCompletion = requireCompletion;
+                cashedRequireCompletion && setRequireCompletion(false);
+                saveData(undefined, false, () => {
+                    cashedRequireCompletion && setRequireCompletion(cashedRequireCompletion);
+                });}}
+            variant="outlined"
+            >
+            Stay on this page
+          </Button>
+          <Button
+            onClick={() => {
+                // Redirect the user to the /
+                // ...but only after the Form has been saved and checked-in
+                saveDataWithCheckin(undefined, () => {
+                    removeWindowHandlers && removeWindowHandlers();
+                    props.history.push("/");
+                });}}
+            variant="contained"
+            color="primary"
+            >
+            Save and exit
+          </Button>
+        </DialogActions>
+      </ResponsiveDialog>
```
I would like this to be in a separate component called "SessionExpiryWarningModal", with the `onClick` handlers as props.

marta- reviewed (dismissed):
Some quick comments, did not test

marta- commented:
@veronikaslc please create a test branch where the session expiry dialog is shown after 1min on inactivity instead of 27, for easier testing.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -482,6 +536,7 @@ function Form (props) {
           ref={formNode}
           className={classNames?.join(' ')}
       >
+      {showSessionExpiredAlert && <Alert severity="error">Your session expired, please refresh to keep editing</Alert>}
```
The specs say "an alert **modal**", so please display this in a dialog. If the page was scrolled to the bottom, the user won't even notice this alert at the top of the form.

Also, it would be more user friendly to be able to click on the word "refresh" or on a "Refresh" button in this dialog to reload the page.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -162,7 +171,41 @@ function Form (props) {
+  }, [isEdit, lastSaveTimestamp, saveInProgress]);
+
+  // display a modal alert that the session will expire in 2 minutes
+  useEffect(() => {
+    if (!isEdit) return;
+    if (!showCountdownModal && countDown > 0 && countDown < 3 * 60 * 1000) {
+      setShowCountdownModal(true);
+    }
+    // After 2 minutes pass and the countdown reaches less than a second time interval
+    if (showCountdownModal && countDown < 1000) {
+      // Do a save & checkin for the current form
+      saveDataWithCheckin(undefined, () => {
+        // On success, display an alert modal
+        setShowCountdownModal(false);
+        setShowSessionExpiredAlert(true);
+      });
+    }
+  }, [countDown]);
```
The implementation does not follow the specs.

In the specs: if inactive, wait for 27 minutes, then show countdown dialog for 2 minutes, then save & checkin at minute 29

In the implementation: if inactive, start counter for 27 minutes, but when we're 3min away from the end (so at min 24) we show the countdown dialog which stays on the screen for 3 minutes, then save and checkin at minute 27.

veronikaslc replied:
Right, initial countDown should be initialised for 30 mins, not 27.

veronikaslc committed:
CARDS-2332: Add automatic saving and checkin of forms after a period of inactivity
Addressed codereview comments

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -21,6 +21,7 @@ import React, { useEffect, useState, useContext } from "react";
 import { withRouter } from "react-router-dom";
 
 import {
+  Alert,
```
Unused, please remove

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -57,6 +58,8 @@ import FormattedText from "../components/FormattedText.jsx";
 import { getFirstIncompleteQuestionEl } from "./FormUtilities.jsx";
 import { getEntityIdentifier } from "../themePage/EntityIdentifier.jsx";
+import SessionExpiryWarningModal from "./SessionExpiryWarningModal.jsx";
+import SessionExpiredAlertModal from "./SessionExpiredAlertModal.jsx";
```
These 2 are never meant to be used separately. Can `SessionExpiredAlertModal` be absorbed into `SessionExpiryWarningModal`?

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -162,7 +171,43 @@ function Form (props) {
+	  setShowCountdownModal(true);
+
+      // start a countdown recalculated every second
+	  let interval = setInterval(() => {
+	      let timeLeft = countDownStartTime + countDown - new Date().getTime();
+	      if (timeLeft >= 0) {
+	        setCountDown(timeLeft);
+	        //  save & checkin at minute 29
+		    if (timeLeft < 1000) {
+			  setShowCountdownModal(false);
+			  clearTimeout(timer);
+			  clearInterval(interval);
+		      // Do a save & checkin for the current form
+		      saveDataWithCheckin(undefined, () => {
+		        // On success, display an alert modal
+		        setShowSessionExpiredAlert(true);
+		      });
+		    }
+	      }
+	    }, 1000);
+	    setAutoCheckinTimer(interval);
```
Can this countdown timer be moved into `SessionExpiryWarningModal`? It is technically "session expiry" functionality, not Form component functionality.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/SessionExpiryWarningModal.jsx:
```diff
@@ -0,0 +1,84 @@
+    diffString("minutes", diffStrings, minutes);
+    diffString("seconds", diffStrings, seconds);
+
+    if (diffStrings.length > 0) {
+      result = "Your session will expire in " + diffStrings[0] + ". You should save your answers now to keep your session active and prevent unsaved data loss.";
+    }
+
+    return result;
+  }
+
+  return (
+      <ResponsiveDialog title="Session will expire soon" open={open}>
```
I think the dialog would look better if the title was the dynamic one "Your session will expire in \<countdown\>" and the message was static "You should save your answers now to keep your session active and prevent data loss." (note also the removal of the word "unsaved" before "data loss"). I will update the specs in Jira.

marta- commented:
@veronikaslc I see that in the test branch the session expiry dialog is set to appear after 1 second. Let's give it at least 30 or 1min so the tester has time to enter some data and test that it's autosaved.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -618,6 +666,27 @@ function Form (props) {
+                setShowCountdownModal(false);
+                // trigger a normal save
+                // ...make sure that on paginated forms this doesn’t go to the next page
+                const cashedRequireCompletion = requireCompletion;
+                cashedRequireCompletion && setRequireCompletion(false);
+                saveData(undefined, false, () => {
+                    cashedRequireCompletion && setRequireCompletion(cashedRequireCompletion);
+                });}}
```
I'm not understanding the purpose of this temporary change to `requireCompletion`. I tested with and without this change and didn't see any difference. What am I missing @veronikaslc ?

veronikaslc replied:
To make sure that on paginated forms this doesn’t go to the next page.

marta- replied:
Can you provide the steps that would make it go to the next page without it? When I fill out the first page, wait for the session expiring dialog, and click on "Stay", I stay on the first page with or without those lines.

veronikaslc replied:
This way we will stop advancing even if we change anything in future, as it was directed by specs `make sure that on paginated forms this doesn’t go to the next page`

marta- replied:
Sorry, I'm not following, not sure what you mean by "if we change anything in the future".

The thing is that here requireCompletion is set to `false`, which means it does _not_ block moving on to the next page (`true` would block). This code likely has no effect.

Again, If I'm missing/misunderstanding something please let me know. Also if you have steps to make the form go to the next page without these lines.

veronikaslc replied:
I'll just remove the code then, indeed.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/SessionExpiryWarningModal.jsx:
```diff
@@ -0,0 +1,84 @@
+        <DialogActions>
+          <Button
+            onClick={() => onStay()}
+            variant="outlined"
```
"Stay" should be the primary action
```suggestion
            variant="contained"
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/SessionExpiryWarningModal.jsx:
```diff
@@ -0,0 +1,84 @@
+          </Button>
+          <Button
+            onClick={() => onExit()}
+            variant="contained"
```
"Leave" should be the secondary action
```suggestion
            variant="outlined"
```

marta- reviewed (commented):
Regression:
* fill out entirely the first page of a paginated form with the `requireCompletion` flag on, e.g. Inpatient survey in prems
* while on the first page (with all required answers filled in), wait for the session expiring dialog to appear
* click "Stay on this page"
* click "Next"
* problem: on the second page the first question is outlined in red
  * this does not happen in dev, and hints to a double save happening while on the second page
  * it also doesn't happen if you move on to the next page before the session expiring dialog appears

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -257,6 +302,9 @@ function Form (props) {
         }
         onSuccess?.();
+        // Every time the user saves, restart the countdown timer
+        autoCheckinTimer && clearInterval(autoCheckinTimer);
+        setCountDown(SESSION_EXPIRATION_DELAY);
```
Can we deal with the autocheckin in a useEffect listening to lastSaveStatus instead of making the save function more complex?

veronikaslc committed:
CARDS-2332: Add automatic saving and checkin of forms after a period of inactivity
Code refactoring

veronikaslc commented:
> * while on the first page (with all required answers filled in), wait for the session expiring dialog to appear
> * click "Stay on this page"
> * click "Next"

@marta- Do you see a double-save request in Networks tab after you click Next? I don't observe it in my build in Chrome. Is it browser-related?

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -618,6 +619,27 @@ function Form (props) {
           <Typography variant="body1" paragraph>Time of the last successful save: {DateTime.fromISO(lastSaveTimestamp.toISOString()).toRelativeCalendar()}</Typography>
         }
       </ErrorDialog>
+      <SessionExpiryWarningModal
```
We should only render the `SessionExpiryWarningModal` in edit mode

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -618,6 +619,27 @@ function Form (props) {
+      <SessionExpiryWarningModal
+        isEdit={isEdit}
+        lastSaveTimestamp={lastSaveTimestamp}
+        saveInProgress={saveInProgress}
```
In the `SessionExpiryWarningModal`, we are only using these 3 once, in the combination `!isEdit || lastSaveTimestamp && saveInProgress`. There's no reason to send all 3 as props if we're going to merge them into 1 for single use. Also, per previous comment, `isEdit` should be guaranteed to be true. Let's pass only one prop combining `lastSaveTimestamp` and `saveInProgress`.

veronikaslc replied:
In this case we indeed could avoid dependency on `isEdit` or `saveInProgress`.
However in general when `useEffect()` could be triggered by each of the `n` variables separately, merging them in one boolean disables functionality. The passed one merged value would not change, hence the hook would not be triggered.

marta- replied:
Once they're passed on as a prop, the effect on that prop will be triggered. The receiving component doesn't care where the prop value is coming from. If the value changes, the effect on it gets executed.

veronikaslc replied:
Now I see what you mean by `merge` - unite then them in one `Object`. I understood merge as passing them as one pre-computed boolean.

veronikaslc committed:
CARDS-2332: Add automatic saving and checkin of forms after a period of inactivity
Code optimisation

marta- commented:
> > * while on the first page (with all required answers filled in), wait for the session expiring dialog to appear
> > * click "Stay on this page"
> > * click "Next"
> 
> @marta- Do you see a double-save request in Networks tab after you click Next? I don't observe it in my build in Chrome. Is it browser-related? I can reproduce the above issue but do not see double-save.

No double save indeed. It's most likely the effect on lastSaveTimestamp that recomputes the incompleteEl and causes it to be highlighted.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -618,6 +619,25 @@ function Form (props) {
+                            // ...but only after the Form has been saved and checked-in
+                            saveDataWithCheckin(undefined, () => {
+                                removeWindowHandlers && removeWindowHandlers();
+                                props.history.push("/");
```
Wouldn't simply navigating away trigger a save and the removal of the window handlers?

veronikaslc replied:
As far as I see on line 347 that step is required.

marta- replied:
I'm pretty sure this line takes care of that functionality: https://github.com/data-team-uhn/cards/blame/dev/modules/homepage/src/main/frontend/src/themePage/index.jsx#L185

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -618,6 +619,25 @@ function Form (props) {
+                            });}}
+                     onExit={() => {
+                            // Redirect the user to the /
+                            // ...but only after the Form has been saved and checked-in
+                            saveDataWithCheckin(undefined, () => {
+                                removeWindowHandlers && removeWindowHandlers();
+                                props.history.push("/");
+                            });}}
```
```suggestion
                           // Navigating away will trigger a save with checkin
                            props.history.push("/");
                            }}
```

veronikaslc committed:
CARDS-2332: Add automatic saving and checkin of forms after a period of inactivity
Removed unnecessary code

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/SessionExpiryWarningModal.jsx:
```diff
@@ -0,0 +1,146 @@
+    diffString("seconds", diffStrings, seconds);
+
+    if (diffStrings.length > 0) {
+      result = "Your session will expire in " + diffStrings[0] + ".";
```
Let's show both minutes (if not 0) and seconds. Also remove the "." at the end of the title. Hopefully this works:
```suggestion
      result = "Your session will expire in " + diffStrings.join(' ');
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/SessionExpiryWarningModal.jsx:
```diff
@@ -0,0 +1,146 @@
+      </ResponsiveDialog>
+      <ResponsiveDialog title="Session expired" open={showSessionExpiredAlert}>
+        <DialogContent dividers>
+          <Typography variant="body1" paragraph>Your session has expired, please refresh to keep editing</Typography>
```
```suggestion
          <Typography variant="body1" paragraph>Your session has expired. Please refresh this page to keep editing.</Typography>
```

marta- commented:
> > > * while on the first page (with all required answers filled in), wait for the session expiring dialog to appear
> > > * click "Stay on this page"
> > > * click "Next"
> > 
> > 
> > @marta- Do you see a double-save request in Networks tab after you click Next? I don't observe it in my build in Chrome. Is it browser-related? I can reproduce the above issue but do not see double-save.
> 
> No double save indeed. It's most likely the effect on lastSaveTimestamp that recomputes the incompleteEl and causes it to be highlighted.

Yes that was it, committing a fix soon.

veronikaslc committed:
CARDS-2332: Add automatic saving and checkin of forms after a period of inactivity

Code optimisation and messae improvements

Co-authored-by: Marta Girdea <marta@phenotips.org>

veronikaslc committed:
CARDS-2332: Add automatic saving and checkin of forms after a period of inactivity

Message improvements

Co-authored-by: Marta Girdea <marta@phenotips.org>

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -618,6 +619,16 @@ function Form (props) {
           <Typography variant="body1" paragraph>Time of the last successful save: {DateTime.fromISO(lastSaveTimestamp.toISOString()).toRelativeCalendar()}</Typography>
         }
       </ErrorDialog>
+      { isEdit && <SessionExpiryWarningModal
+                    lastSaveTimestamp={lastSaveTimestamp}
+                    saveDataWithCheckin={saveDataWithCheckin}
+                    onStay={() => saveData(undefined, false)}
+                    onExit={() => {
+                            // Redirect the user to the /
+                           // Navigating away will trigger a save with checkin
+                            props.history.push("/");
+                            }}
+                    /> }
```
How about some indentation cleanup? Also, I don't think we need those comments.

```suggestion
      { isEdit &&
        <SessionExpiryWarningModal
          lastSaveTimestamp={lastSaveTimestamp}
          saveDataWithCheckin={saveDataWithCheckin}
          onStay={() => saveData(undefined, false)}
          onExit={() => props.history.push("/")}
        />
      }
```

marta- replied:
Hm, the error dialog just above should also be rendered only in edit mode...

marta- commented:
> > > > * while on the first page (with all required answers filled in), wait for the session expiring dialog to appear
> > > > * click "Stay on this page"
> > > > * click "Next"
> > > 
> > > 
> > > @marta- Do you see a double-save request in Networks tab after you click Next? I don't observe it in my build in Chrome. Is it browser-related? I can reproduce the above issue but do not see double-save.
> > 
> > 
> > No double save indeed. It's most likely the effect on lastSaveTimestamp that recomputes the incompleteEl and causes it to be highlighted.
> 
> Yes that was it, committing a fix soon.

False alarm, my fix isn't working (yet).

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/SessionExpiryWarningModal.jsx:
```diff
@@ -0,0 +1,146 @@
+ * @param {func} saveDataWithCheckin Callback for when to save and checkin form after countdown is timedout
+ */
+function SessionExpiryWarningModal(props) {
+  const { lastSaveTimestamp, onStay, onExit, saveDataWithCheckin } = props;
```
I would rename the `saveDataWithCheckin` prop to `checkinHandler`, in case we want to use `SessionExpiryWarningModal` for checking in other types of resources in the future.

marta- replied:
On second thought, I would not even pass a handler.  The session will expire whether the save is successful or not and we need to display the "Session expired" alert regardless.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/SessionExpiryWarningModal.jsx:
```diff
@@ -0,0 +1,146 @@
+  let [ showSessionExpiredAlert, setShowSessionExpiredAlert ] = useState(false);
+  let [ autoCheckinTimer, setAutoCheckinTimer ] = useState();
+  const SESSION_EXPIRATION_DELAY = 2 * 60 * 1000;
+  let [ countDown, setCountDown ] = useState(SESSION_EXPIRATION_DELAY);
```
I would like the displayed countdown to start at 1min59 instead of 2min and I'd like to see 0s before the counter disappears.

Also, can anyone confirm that it jumps from 2min to 1min58? It happens to me consistently @sashaandjic @veronikaslc

sashaandjic replied:
yes. it starts with `2 mins` for 1 sec then it turns into `1 min` for 58 sec, then it turns into seconds countdown from 59 to 1. the whole thing lasts 1min + 59 sec rather than 2 minutes.

VIDEO
-----
https://github.com/data-team-uhn/cards/assets/23243244/d892ade9-690f-4fa0-8d0e-1b3c7eb8c736

SCREENSHOTS
------

The `1 min` shows up after 1 second
![image](https://github.com/data-team-uhn/cards/assets/23243244/20041276-03fb-4cb3-8bd8-0a18dd221a69)

`1 min` lasts for 58 seconds, rather than 59 seconds
![image](https://github.com/data-team-uhn/cards/assets/23243244/47ed9444-505d-400b-a734-c00a27b06d4d)

marta- replied:
Sorry @sashaandjic , looks like you're looking at the test branch which is now outdated. I'll update it when I get the chance.

marta- replied:
This small issue no longer occurs after commit 4a9d111

marta- committed:
CARDS-2332: Add automatic saving and checkin of forms after a period of inactivity
Bugfix: don't highlight incomplete required questions when autosaving

marta- committed:
CARDS-2332: Add automatic saving and checkin of forms after a period of inactivity
Indentation cleanup

marta- commented:
**Bug** - the countdown is not properly reset:
* open a form for editing
* wait for the session expiring to appear
* let it count down for about 10s
* click "Stay on this page"
* wait until it shows up again and note how the coundown jumps from "2 minutes" to "1 minute 49 seconds"
* let it count down for another 10s, click stay, wait until the dialog appears again and note how  the coundown jumps from "2 minutes" to "1 minute 39 seconds"
* ... and so on

marta- commented:
> Regression:
> 
> * fill out entirely the first page of a paginated form with the `requireCompletion` flag on, e.g. Inpatient survey in prems
> * while on the first page (with all required answers filled in), wait for the session expiring dialog to appear
> * click "Stay on this page"
> * click "Next"
> * problem: on the second page the first question is outlined in red
>   
>   * this does not happen in dev, and hints to a double save happening while on the second page
>   * it also doesn't happen if you move on to the next page before the session expiring dialog appears

Fixed in 4c4bffd

marta- committed:
CARDS-2332: Add automatic saving and checkin of forms after a period of inactivity
Better props for the SessionExpiryWarningModal component, improved autosave settings

marta- committed:
CARDS-2332: Add automatic saving and checkin of forms after a period of inactivity
* Bugfix: countdown timer not reset after the first launch
* UI fix: minor flickering when switching from countdown to "session expired"
* Extensibility improvement: made the expiry and countdown configurable, defaulting to 29min and 2min
* Refactoring: clearer variable names

marta- commented:
> **Bug** - the countdown is not properly reset:
> 
> * open a form for editing
> * wait for the session expiring to appear
> * let it count down for about 10s
> * click "Stay on this page"
> * wait until it shows up again and note how the coundown jumps from "2 minutes" to "1 minute 49 seconds"
> * let it count down for another 10s, click stay, wait until the dialog appears again and note how  the coundown jumps from "2 minutes" to "1 minute 39 seconds"
> * ... and so on

Fixed in 4a9d111

marta- commented:
@sashaandjic Ready for testing using branch `CARDS-2332-test` where the Session expiring warning appears after 30s instead of 27min. Patient UI needs to be thoroughly tested as well (I personally haven't looked at it yet).

sashaandjic commented:
When patient hits `SAVE AND EXIT`, the patient should be back at the welcome screen rather than blank screen.

https://github.com/data-team-uhn/cards/assets/23243244/2df4300d-2b0a-4630-9f2b-d5cdc2c03083

marta- commented:
> When patient hits `SAVE AND EXIT`, the patient should be back at the welcome screen rather than blank screen.

That's a logic/specs bug. Not sure what options we have to approach the patient vs non-patient behavior elegantly. Do we reload if there's an auth_token and go to / otherwise? @sdumitriu please advise.

sdumitriu commented:
> > When patient hits `SAVE AND EXIT`, the patient should be back at the welcome screen rather than blank screen.
> 
> That's a logic/specs bug. Not sure what options we have to approach the patient vs non-patient behavior elegantly. Do we reload if there's an auth_token and go to / otherwise? @sdumitriu please advise.

I think that loading `/` should have the desired effect in production. As a patient, that actually loads https://yourexperience.uhn.ca/Survey.html which displays the `To fill out surveys, please follow the personalized link that was emailed to you` message. As a clinician, it will display the dashboard.

marta- commented:
> > > When patient hits `SAVE AND EXIT`, the patient should be back at the welcome screen rather than blank screen.
> > 
> > 
> > That's a logic/specs bug. Not sure what options we have to approach the patient vs non-patient behavior elegantly. Do we reload if there's an auth_token and go to / otherwise? @sdumitriu please advise.
> 
> I think that loading `/` should have the desired effect in production. As a patient, that actually loads https://yourexperience.uhn.ca/Survey.html which displays the `To fill out surveys, please follow the personalized link that was emailed to you` message. As a clinician, it will display the dashboard.

What about in data-pro? Reminder that the main reason why we're implementing this is to get forms checked in when abandoned, so we can implement the epic pull efficiently in data-pro.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -618,6 +634,14 @@ function Form (props) {
+        <SessionExpiryWarningModal
+          lastActivityTimestamp={lastSaveTimestamp}
+          onStay={() => setAutosaveOptions({})}
+          onExit={() => props.history.push("/")}
```
According to @sdumitriu , manipulating window.location instead of doing history.push here should prevent the blank screen for patients.

marta- commented:
> When patient hits `SAVE AND EXIT`, the patient should be back at the welcome screen rather than blank screen.

@veronikaslc : Per @sdumitriu 's suggestion, we should try adding a route from `/` (exact match) to `Survey.html` in the [Patient Portal index](https://github.com/data-team-uhn/cards/blob/dev/modules/patient-portal/src/main/frontend/src/patient-portal/index.jsx#L100-L103). We need to test to make sure this doesn't break any routing for non-patient users.

veronikaslc committed:
CARDS-2332: Add automatic saving and checkin of forms after a period of inactivity
Added a route from / (exact match) to Survey.html

sashaandjic reviewed (approved):
tested - all good

sdumitriu reviewed (commented):
The "Session expired" modal should have a fully opaque background to hide the data.

Is it easy to disable the "This page is asking you to confirm that you want to leave — information you’ve entered may not be saved" alert when the session expired and the Refresh button is pressed?

veronikaslc committed:
CARDS-2332: Add automatic saving and checkin of forms after a period of inactivity
Prevent the default window confirmation dialog on user leaving the page

veronikaslc commented:
> The "Session expired" modal should have a fully opaque background to hide the data.

Now the background is rgba(0, 0, 0, 0.5). What is the desired style?

marta- commented:
> > The "Session expired" modal should have a fully opaque background to hide the data.
> 
> Now the background is rgba(0, 0, 0, 0.5). What is the desired style?

When `expired` becomes true, I would render a 2nd `Backdrop` just behind the dialog, with bg color `theme.palette.background.paper` (which is white) to make the data "disappear" without a huge shift in the perceived background color.

veronikaslc committed:
CARDS-2332: Add automatic saving and checkin of forms after a period of inactivity
Added opaque Backdrop to the session expired dialog state

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/SessionExpiryWarningModal.jsx:
```diff
@@ -0,0 +1,179 @@
+    !open && countdownTimer && clearInterval(countdownTimer);
+  }, [open, countdownTimer]);
+
+  // prevent the default window confirmation dialog on user leaving the page
+  useEffect(() => {
+	open && window.addEventListener("beforeunload", (event) => event.preventDefault());
+    return (() => {
+      window.removeEventListener("beforeunload", (event) => event.preventDefault());
+    });
```
This has the exact opposite effect, calling `preventDefault` is how to trigger the confirmation dialog, and this is already done in [Form.jsx](https://github.com/data-team-uhn/cards/blob/cb4579ab09884f72302c6eef3ab2c570a3fbbcd9/modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx#L229). And it's not just the confirmation that shows up, it's also another save request triggered in the background. And that save request will actually cause the login dialog to show up. What needs to happen is the complete removal of the event handler, and it seems that there is already an easy way to do that: pass `removeWindowHandlers` to the dialog and call it when the page expired.

sdumitriu replied:
Actually it doesn't even have to be passed, we already have the `onExpired` callback, we can call it from there.

veronikaslc replied:
@sdumitriu I made the change you requested but I'm not sure how calling onExpired from there if you mean from `useEffect()` suppose to work. Also we already call `onExpired` when countdown runs to 0, when the session actually expires, how come calling it second time from anywhere makes a difference?

sdumitriu replied:
I just pushed a fix for this.

It's not calling about `onExpired` again, but to call `removeWindowHandlers` within `onExpired`.

There was another bug which was preventing this solution from working, `removeWindowHandlers` wasn't removing the handler correctly.

veronikaslc replied:
Ok, now I understand what you mean!

veronikaslc committed:
CARDS-2332: Add automatic saving and checkin of forms after a period of inactivity

Prevent the default window confirmation dialog on user leaving the page

sdumitriu committed:
CARDS-2332: Add automatic saving and checkin of forms after a period of inactivity
Prevent the default window confirmation dialog on user leaving the page

sdumitriu reviewed (approved):
Works well enough for now.

@marta- I tried automatically loading an expired page by using the `onSuccess` setting of `autosaveOptions`, and while that worked perfectly, I'm not sure it's the right thing to do. On one hand, for privacy it's better, because it also hides the URL from before, which means that if a patient has their auth token in the URL on a shared device, the next person to click on refresh will not get authenticated as the previous user. But this is a bad fix, the auth token is still one Back button away. On the other hand, if the clinician user was doing something and left for lunch, it would be nice to be able to jump right in without having to remember what patient they were working on, and finding the right form to edit.

marta- commented:
> On one hand, for privacy it's better, because it also hides the URL from before, which means that if a patient has their auth token in the URL on a shared device, the next person to click on refresh will not get authenticated as the previous user. But this is a bad fix, the auth token is still one Back button away.

`history.replace` instead of `history.push` should take care of that.

> On the other hand, if the clinician user was doing something and left for lunch, it would be nice to be able to jump right in without having to remember what patient they were working on, and finding the right form to edit.

As I said before, I would like us to have an elegant way to define patient vs clinician workflows. I don't think now in this PR is the time to make that improvement, though.

sashaandjic reviewed (approved):
tested
